app/src/main/java/com/fivegmag/a5gmsmediasessionhandler/MediaSessionHandlerMessengerService.kt

```diff
@@ -302,5 +323,177 @@ class MediaSessionHandlerMessengerService() : Service() {
         return mMessenger.binder
     }
 
+    /** If the consumption reporting procedure is activated. Refer to TS26.512 Clause 4.7.4
+     * When the clientConsumptionReportingConfiguration.samplePercentage value is 100, the Media Session Handler shall activate the consumption reporting procedure.
+     * If the samplePercentage is less than 100, the Media Session Handler shall generate a random number which is uniformly distributed in the range of 0 to 100,
+     * and the Media Session Handler shall activate the consumption report procedure when the generated random number is of a lower value than the samplePercentage value.
+     */
+    private fun isConsumptionReportingActivated(clientId: Int): Boolean {
+        //if(currentServiceAccessInformation.isInitialized)
+        if(clientsSessionData[clientId]?.serviceAccessInformation == null) {
+            Log.i(TAG, "[ConsumptionReporting] IsConsumptionReportingActivated: ServiceAccessInformation is 【NULL】")
+            return false
+        }
+
+        // check if the samplePercentage in ServiceAccessInformation.clientConsumptionReportingConfiguration is valid
+        var samplePercentage: Float =  clientsSessionData[clientId]?.serviceAccessInformation!!.clientConsumptionReportingConfiguration.samplePercentage;
+        if(samplePercentage > SamplePercentageMax || samplePercentage < 0)
+        {
+            Log.i(TAG, "[ConsumptionReporting] Invaild samplePercentage[$samplePercentage] in ServiceAccessInformation.clientConsumptionReportingConfiguration")
+            return false;
+        }
+
+        // if samplePercentage is 100, MSH shall activate the consumption reporting procedure
+        if(abs(SamplePercentageMax - samplePercentage) < EPSILON)
+        {
+            Log.i(TAG, "[ConsumptionReporting] SamplePercentage==SamplePercentageMax, always report")
+            return true;
+        }
+
+        // if the generated random number is of a lower value than the samplePercentage value
```

Comment:
This seems to be re-evaluated for each consumption report while it should only be re-evaluated for each new streaming session

Reply:
As I understand, it's purpose is to control the percentage of report(to reduce network footprint?)
I think may the description of this part of the spec is very unclear.@rjb1000

Reply:
Yes, you're right, @shilinding. When less than 100% of 5GMS Clients send consumption reports, the load on the 5GMS AF is reduced.

TS 26.512 clause 11.2.3.1 defines _ServiceAccessInformaion.clientConsumptionReportingConfiguration.samplePercentage_ as:

> The percentage of media streaming sessions that shall send consumption reports, expressed as a floating point value between 0.0 and 100.0.

When we discussed how to implement this aspect of metrics reporting, we realised that there is an implementation choice about exactly when to evaluate the random number generator that drive this behaviour:

1. Generate the random number once for each streaming session. Reporting is then either enabled or disabled for that streaming session.
2. Generate the random number each time you are thinking about sending a report.

This is left ambiguous in TS 26.512, and I think that's fine because it is a case where we can leave it to implementation choice and get a similar server load result across an entire population of 5GMS Clients.

For metrics reporting, we chose option 1. I think @dsilhavy's comment is suggesting that we choose option 1 for consumption reporting as well for consistency.

For option 1, there remains an open question in my mind about the implementation. If the random number determines that reporting is disabled for a session, does the 5GMS Client still need to _collect_ samples, just in case the reporting configuration changes? Should our 5GMS Client implementation re-evaluate the random number if the client configuration changes? And, if it then needs to start reporting, does it need to report historic data? (Maybe that applies more to metrics reports than consumption reports, though.)

Reply:
Probably related to the issue created here: https://github.com/5G-MAG/rt-5gms-media-session-handler/issues/32